# Case 2: Ranking in Sports
### Operations Research

In [1]:
#imports
import numpy as np
import pandas as pd
import os

## A. Collect Data
In our report, we chose a historical data set for football. We collected the data set of the Eredivisie from year 2008 until 2022. All years are complete, except 2020 due to the Covid-crisis. The 2022 season is currently ongoing, so this season is not fully completed.

In [2]:
data_files = os.listdir("data")
data = pd.read_csv(os.path.join("data", data_files[0]))

for data_file in data_files[1:]:
    data = pd.concat([data, pd.read_csv(os.path.join("data", data_file))])
    
# Remove whitespaces
data.loc[:, "HomeTeam"] = data["HomeTeam"].str.strip()
data.loc[:, "AwayTeam"] = data["AwayTeam"].str.strip()

#Some teams changed their names
data = data.replace("Sparta", "Sparta Rotterdam")
data = data.replace("Roda", "Roda JC")

data

Div      Date   HomeTeam       AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0    N1  29/08/08    Vitesse      Groningen   0.0   4.0   A   0.0   1.0   A   
1    N1  30/08/08   Nijmegen     Graafschap   2.0   0.0   H   0.0   0.0   D   
2    N1  30/08/08    Roda JC         Twente   1.0   1.0   D   0.0   0.0   D   
3    N1  30/08/08    Utrecht  PSV Eindhoven   1.0   5.0   A   0.0   3.0   A   
4    N1  30/08/08  Willem II           Ajax   2.0   1.0   H   0.0   1.0   A   
..   ..       ...        ...            ...   ...   ...  ..   ...   ...  ..   
301  N1  06/05/12   Heracles       Nijmegen   1.0   2.0   A   0.0   1.0   A   
302  N1  06/05/12  NAC Breda       Waalwijk   3.0   2.0   H   1.0   1.0   D   
303  N1  06/05/12    Roda JC        Utrecht   1.0   3.0   A   0.0   1.0   A   
304  N1  06/05/12    Vitesse           Ajax   1.0   3.0   A   1.0   1.0   D   
305  N1  06/05/12  VVV Venlo         Twente   4.0   2.0   H   1.0   1.0   D   

     ...  B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  \
0    ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
1    ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
2    ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
3    ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
4    ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
..   ...       ...    ...    ...      ...      ...      ...      ...   
301  ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
302  ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
303  ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
304  ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   
305  ...       NaN    NaN    NaN      NaN      NaN      NaN      NaN   

     Unnamed: 58  Unnamed: 59  Unnamed: 60  
0            NaN          NaN          NaN  
1            NaN          NaN          NaN  
2            NaN          NaN          NaN  
3            NaN          NaN          NaN  
4            NaN          NaN          NaN  
..           ...          ...          ...  
301          NaN          NaN          NaN  
302          NaN          NaN          NaN  
303          NaN          NaN          NaN  
304          NaN          NaN          NaN  
305          NaN          NaN          NaN  

[4113 rows x 141 columns]

## B. Construct the Graph
First, we pick the most important variables for every match. To get a general overview of every team, we may have to change the home and away team. This will result in a home team that is before in the alphabet. For example: Excelsior - Cambuur will be changed into Cambuur - Excelsior. 

In [3]:
scores = data.loc[:, ["HomeTeam", "AwayTeam", "FTHG", "FTAG", "FTR", "Date"]].rename({"FTHG": "HomeGoals", "FTAG": "AwayGoals", "FTR": "HomeWin"}, axis=1)
scores["AwayWin"] = scores["HomeWin"] == "A"
scores["HomeWin"] = scores["HomeWin"] == "H"
scores["Draw"] = (scores["HomeWin"] == False) & (scores["AwayWin"] == False)

scores = scores.dropna()

scores.loc[scores["HomeTeam"] > scores["AwayTeam"]] = scores.loc[scores["HomeTeam"] > scores["AwayTeam"]].rename({"HomeTeam": "AwayTeam", "AwayTeam": "HomeTeam", "AwayGoals": "HomeGoals", "HomeGoals": "AwayGoals", "HomeWin": "AwayWin", "AwayWin": "HomeWin"}, axis=1)
scores.set_index(["HomeTeam", "AwayTeam"], inplace=True)
scores = scores.rename({"HomeGoals": "FirstGoals", "AwayGoals": "SecondGoals", "HomeWin": "FirstWin", "AwayWin": "SecondWin"}, axis=1)
scores

FirstGoals  SecondGoals  FirstWin      Date  \
HomeTeam      AwayTeam                                                 
Groningen     Vitesse           4.0          0.0      True  29/08/08   
Graafschap    Nijmegen          0.0          2.0     False  30/08/08   
Roda JC       Twente            1.0          1.0     False  30/08/08   
PSV Eindhoven Utrecht           5.0          1.0      True  30/08/08   
Ajax          Willem II         1.0          2.0     False  30/08/08   
...                             ...          ...       ...       ...   
Heracles      Nijmegen          1.0          2.0     False  06/05/12   
NAC Breda     Waalwijk          3.0          2.0      True  06/05/12   
Roda JC       Utrecht           1.0          3.0     False  06/05/12   
Ajax          Vitesse           3.0          1.0      True  06/05/12   
Twente        VVV Venlo         2.0          4.0     False  06/05/12   

                         SecondWin   Draw  
HomeTeam      AwayTeam                     
Groningen     Vitesse        False  False  
Graafschap    Nijmegen        True  False  
Roda JC       Twente         False   True  
PSV Eindhoven Utrecht        False  False  
Ajax          Willem II       True  False  
...                            ...    ...  
Heracles      Nijmegen        True  False  
NAC Breda     Waalwijk       False  False  
Roda JC       Utrecht         True  False  
Ajax          Vitesse        False  False  
Twente        VVV Venlo       True  False  

[4110 rows x 6 columns]

Every match is sorted, so we can continue to combine these matches. The most important statistics are added up for each team.

In [4]:
stats = scores.groupby(scores.index).sum()
number_of_games = scores.groupby(scores.index).size().to_frame("GamesNumber")
stats = stats.merge(number_of_games, left_index=True, right_index=True)
stats.index = pd.MultiIndex.from_tuples(stats.index)
stats

FirstGoals  SecondGoals  FirstWin  SecondWin  Draw  \
AZ Alkmaar Ajax             30.0         56.0         8         14     5   
           Cambuur          14.0          5.0         5          0     3   
           Den Haag         58.0         23.0        18          5     3   
           Dordrecht         5.0          1.0         2          0     0   
           Excelsior        25.0         21.0         5          3     6   
...                          ...          ...       ...        ...   ...   
Vitesse    Zwolle           40.0         20.0        14          4     2   
Volendam   Willem II         4.0          3.0         1          0     1   
Waalwijk   Willem II         5.0         13.0         0          7     2   
           Zwolle            8.0         13.0         0          2     7   
Willem II  Zwolle           16.0         21.0         6          9     3   

                      GamesNumber  
AZ Alkmaar Ajax                27  
           Cambuur              8  
           Den Haag            26  
           Dordrecht            2  
           Excelsior           14  
...                           ...  
Vitesse    Zwolle              20  
Volendam   Willem II            2  
Waalwijk   Willem II            9  
           Zwolle               9  
Willem II  Zwolle              18  

[326 rows x 6 columns]

We want to make a weighted directed graph, but first we have to compute the weights between teams. We can define the weights as follows:
$f_{i, j} = \frac{l_{i, j} + 0.5d_{i, j}}{2g_{i, j}} + \frac{c_{i, j}}{2c_{i, j} + 2s_{i, j}}$,
where $l_{i, j}$ is games lost, $d_{i, j}$ is games that result in a draw, $g_{i, j}$  is the total games. $c_{i, j}$ is the total goals conceded, $s_{i, j}$ is total goals scored. $f_{i,j}$ is the weigth of the first team $i$ against second team $j$. Note that $f_{i, j} + f_{j, i} = 1$. 

A good performing team has a low weight and and a bad performing team has a high weight. 

We print a table with the weights. The teams on the left is the first team $i$ and the teams on the top is the second team $j$.

In [5]:
weights_series = (stats["SecondWin"] + 0.5*stats["Draw"])/(2 * stats["GamesNumber"]) + (stats["SecondGoals"]/(2*stats["FirstGoals"] + 2*stats["SecondGoals"]))

weights_copy = weights_series.copy()
second_index = pd.MultiIndex.from_arrays([weights_copy.index.get_level_values(1), weights_copy.index.get_level_values(0)])
weights_copy.index = second_index
weights_copy = 1 - weights_copy

weights = pd.concat([weights_series, weights_copy]).sort_index().unstack()
# weights.replace(np.nan, 0, inplace=True)
weights

AZ Alkmaar      Ajax   Cambuur  Den Haag  Dordrecht  \
AZ Alkmaar               NaN  0.631137  0.225329  0.266975   0.083333   
Ajax                0.368863       NaN  0.119048  0.194919   0.392857   
Cambuur             0.774671  0.880952       NaN  0.633333   0.225000   
Den Haag            0.733025  0.805081  0.366667       NaN   0.125000   
Dordrecht           0.916667  0.607143  0.775000  0.875000        NaN   
Excelsior           0.557453  0.860512  0.522727  0.711823   0.708333   
FC Emmen            0.966667  0.961538       NaN  0.253846        NaN   
Feyenoord           0.447637  0.736003  0.111111  0.407273   0.100000   
For Sittard         0.829670  0.951613  0.000000  0.300000        NaN   
Go Ahead Eagles     0.799107  0.916667  0.834615  0.458333   0.708333   
Graafschap          0.828226  0.916860  0.347222  0.422222        NaN   
Groningen           0.578704  0.750597  0.485119  0.354030   0.250000   
Heerenveen          0.578283  0.796616  0.403571  0.389277   0.708333   
Heracles            0.778912  0.788206  0.539773  0.364937   0.339286   
NAC Breda           0.616667  0.897569  0.162500  0.570513   0.325000   
Nijmegen            0.545194  0.863025  0.644118  0.521723        NaN   
PSV Eindhoven       0.361404  0.569368  0.121921  0.164395   0.071429   
Roda JC             0.641844  0.811153  0.354167  0.460249        NaN   
Sparta Rotterdam    0.764311  0.804010  0.791667  0.418860        NaN   
Twente              0.565000  0.722564  0.309524  0.399972   0.000000   
Utrecht             0.546339  0.576479  0.557112  0.416209   0.090909   
VVV Venlo           0.760000  0.942350       NaN  0.571739        NaN   
Vitesse             0.610324  0.681319  0.240476  0.466418   0.090909   
Volendam            1.000000  0.833333       NaN  1.000000        NaN   
Waalwijk            0.789263  0.939103  0.716667  0.567130        NaN   
Willem II           0.634910  0.816658  0.414286  0.668737   0.071429   
Zwolle              0.751949  0.861842  0.507056  0.372899   0.071429   

                  Excelsior  FC Emmen  Feyenoord  For Sittard  \
AZ Alkmaar         0.442547  0.033333   0.552363     0.170330   
Ajax               0.139488  0.038462   0.263997     0.048387   
Cambuur            0.477273       NaN   0.888889     1.000000   
Den Haag           0.288177  0.746154   0.592727     0.700000   
Dordrecht          0.291667       NaN   0.900000          NaN   
Excelsior               NaN  0.166667   0.748626     0.732143   
FC Emmen           0.833333       NaN   0.776667     0.446429   
Feyenoord          0.251374  0.223333        NaN     0.309524   
For Sittard        0.267857  0.553571   0.690476          NaN   
Go Ahead Eagles    0.450000       NaN   0.741429     0.214286   
Graafschap         0.433036  0.291667   0.669828     0.450000   
Groningen          0.364286  0.362179   0.669683     0.311905   
Heerenveen         0.391337  0.635714   0.641266     0.449519   
Heracles           0.443681  0.395833   0.712963     0.345238   
NAC Breda          0.464130  0.750000   0.711538     0.812500   
Nijmegen           0.443750       NaN   0.641478     0.125000   
PSV Eindhoven      0.128968  0.217647   0.518954     0.062500   
Roda JC            0.270690       NaN   0.743852          NaN   
Sparta Rotterdam   0.731618  0.379808   0.727106     0.400000   
Twente             0.500000  0.366071   0.614087     0.519231   
Utrecht            0.375000  0.566667   0.662827     0.263889   
VVV Venlo          0.547917  0.583333   0.749645     0.642857   
Vitesse            0.376246  0.324561   0.617031     0.335165   
Volendam                NaN       NaN   0.625000          NaN   
Waalwijk           0.500000  0.535714   0.737557     0.627778   
Willem II          0.571078  0.433333   0.773249     0.410714   
Zwolle             0.211111  0.303846   0.658772     0.523810   

                  Go Ahead Eagles  ...   Roda JC  Sparta Rotterdam    Twente  \
AZ Alkmaar               0.200893  ...  0.358156          0.235689  0.43

To determine the best team, we want to make a transition matrix. The matrix combines all $p_{ij}$ values for every $i,j$. Where $p_{ij}=\mathbb{P}(X_{n+1}=j|X_n=i)$. 

This matrix has two important properties. A probability is always greater or equal to zero ($p_{ij}\geq0$). There are some empty values in the matrix, because some teams have not played against each other. To prevent empty cells, we use a value called a damping factor. The report "PageRank Approach to Ranking National Football Teams" by Verica Lazova and Lasko Basnarkov shows that a low dumping factor results in less errors. In this report, we choose a damping factor of $0.05$.

The transition matrix has one other property. The sum of every row in equal to one ($\sum_{j\in S}p_{ij}=1$). So we rescale every value such that the sum of every row is one.

We use the following equation for calculating the elements of transition probability matrix $Q$, with damping factor $d$, the weights matrix $A$ and $N$ the number of clubs. The result is an ergodic matrix.

$Q_{i,j} = (1-d) * \frac{A_{i,j}}{\sum_{k \in N}{A_{i,k}}} + \frac{d}{N} $

In [6]:
damping_factor = 0.05 # geen idee waarom dit, maar is wel prima denk ik
weights_sum_df = weights.sum(axis=1)
sum_of_weights_matrix = np.array([weights_sum_df])
sum_of_weights_df = pd.DataFrame(np.repeat(sum_of_weights_matrix, len(weights_sum_df), axis=0))
sum_of_weights_df.index = list(weights_sum_df.index)
sum_of_weights_df.columns = list(weights_sum_df.index)
number_of_clubs = len(sum_of_weights_df.index)
cleaned_weights = weights.replace(np.nan, 0)
transition_probability_matrix = (1 - damping_factor) * cleaned_weights/sum_of_weights_df.transpose() + damping_factor/number_of_clubs
transition_probability_matrix

AZ Alkmaar      Ajax   Cambuur  Den Haag  Dordrecht  \
AZ Alkmaar          0.001852  0.074485  0.027783  0.032576   0.011442   
Ajax                0.070855  0.001852  0.024122  0.038315   0.075344   
Cambuur             0.056567  0.064074  0.001852  0.046584   0.017744   
Den Haag            0.052427  0.057398  0.027150  0.001852   0.010476   
Dordrecht           0.070756  0.047489  0.060107  0.067624   0.001852   
Excelsior           0.037621  0.057067  0.035393  0.047526   0.047302   
FC Emmen            0.077797  0.077394  0.001852  0.021795   0.001852   
Feyenoord           0.052659  0.085389  0.014463  0.048078   0.013202   
For Sittard         0.062024  0.070868  0.001852  0.023610   0.001852   
Go Ahead Eagles     0.057203  0.065345  0.059662  0.033599   0.050915   
Graafschap          0.053634  0.059176  0.023561  0.028250   0.001852   
Groningen           0.050153  0.064500  0.042342  0.031401   0.022718   
Heerenveen          0.046852  0.063842  0.033257  0.032144   0.056972   
Heracles            0.058648  0.059326  0.041211  0.028462   0.026592   
NAC Breda           0.042589  0.061146  0.012587  0.039540   0.023322   
Nijmegen            0.043400  0.067622  0.050939  0.041611   0.001852   
PSV Eindhoven       0.058500  0.091097  0.020962  0.027620   0.013048   
Roda JC             0.049014  0.061455  0.027876  0.035671   0.001852   
Sparta Rotterdam    0.052227  0.054844  0.054030  0.029459   0.001852   
Twente              0.055319  0.070229  0.031143  0.039702   0.001852   
Utrecht             0.050256  0.052926  0.051210  0.038727   0.009906   
VVV Venlo           0.054673  0.067346  0.001852  0.041588   0.001852   
Vitesse             0.055380  0.061607  0.022943  0.042759   0.009825   
Volendam            0.090384  0.075629  0.001852  0.090384   0.001852   
Waalwijk            0.053077  0.062802  0.048365  0.038660   0.001852   
Willem II           0.043087  0.054891  0.028758  0.045284   0.006491   
Zwolle              0.059430  0.067844  0.040678  0.030405   0.007321   

                  Excelsior  FC Emmen  Feyenoord  For Sittard  \
AZ Alkmaar         0.052781  0.005688   0.065419     0.021454   
Ajax               0.027946  0.009047   0.051238     0.010904   
Cambuur            0.035562  0.001852   0.064634     0.072482   
Den Haag           0.021735  0.053332   0.042747     0.050148   
Dordrecht          0.023776  0.001852   0.069503     0.001852   
Excelsior          0.001852  0.012546   0.049888     0.048830   
FC Emmen           0.067322  0.001852   0.062870     0.036925   
Feyenoord          0.030383  0.027200   0.001852     0.036983   
For Sittard        0.021278  0.042000   0.051929     0.001852   
Go Ahead Eagles    0.033021  0.001852   0.053207     0.016695   
Graafschap         0.028926  0.020087   0.043731     0.029987   
Groningen          0.032257  0.032081   0.057746     0.027885   
Heerenveen         0.032305  0.051321   0.051753     0.036832   
Heracles           0.034204  0.030715   0.053839     0.027026   
NAC Breda          0.032513  0.051398   0.048857     0.055526   
Nijmegen           0.035669  0.001852   0.050738     0.011378   
PSV Eindhoven      0.022067  0.035967   0.083195     0.011648   
Roda JC            0.021742  0.001852   0.056510     0.001852   
Sparta Rotterdam   0.050072  0.026885   0.049775     0.028216   
Twente             0.049168  0.036494   0.059964     0.050987   
Utrecht            0.035076  0.052057   0.060576     0.025232   
VVV Venlo          0.039933  0.042394   0.053953     0.046531   
Vitesse            0.034850  0.030318   0.055969     0.031247   
Volendam           0.001852  0.001852   0.057184     0.001852   
Waalwijk           0.034303  0.036621   0.049721     0.042596   
Willem II          0.038942  0.029995   0.052072     0.028526   
Zwolle             0.018017  0.025118   0.052295     0.041961   

                  Go Ahead Eagles  ...   Roda JC  Sparta Rotterdam    Twente  \
AZ Alkmaar               0.024971  ...  0.043070          0.028976  0.05

## C. Compute the PageRank
This is the state transition matrix. We can apply the PageRank algorithm to rank all teams. The PageRank is calculated using the power method.  The best team has the highest rank.

In [7]:
pagerank = np.array(transition_probability_matrix)

accuracy = 5
while (pagerank.round(accuracy)[0] != pagerank.round(accuracy)[1]).any():
    pagerank =  pagerank @ pagerank

ranking = pd.DataFrame(pagerank[0], index=transition_probability_matrix.index, columns=["pagerank"])
ranking.sort_values(["pagerank"], ascending=False)

pagerank
Ajax              0.062405
PSV Eindhoven     0.059093
Feyenoord         0.053449
AZ Alkmaar        0.052411
Twente            0.048159
Vitesse           0.045514
Utrecht           0.044502
Groningen         0.043022
Heerenveen        0.042815
Heracles          0.038027
Zwolle            0.036814
Den Haag          0.036560
Nijmegen          0.036172
NAC Breda         0.035255
Willem II         0.034226
Roda JC           0.032992
Sparta Rotterdam  0.032847
Excelsior         0.032344
Waalwijk          0.030277
For Sittard       0.029582
Cambuur           0.029218
Graafschap        0.027398
VVV Venlo         0.027045
Go Ahead Eagles   0.026594
FC Emmen          0.026040
Volendam          0.019439
Dordrecht         0.017800

From this PageRank, we can conclude that Ajax is the best team with a PageRank of approximately $0.0624$.